Node affinity scheduling strategy (#23381)
ray-project/ray: doc/source/ray-core/actors/scheduling.rst

Scheduling Strategy
-------------------
Actors support a ``scheduling_strategy`` option to specify the strategy used to decide the best node among available nodes.
Currently the supported strategies for actors are ``"DEFAULT"`` and ``ray.util.scheduling_strategies.NodeAffinitySchedulingStrategy(node_id, soft: bool)``.

``"DEFAULT"`` is the default strategy used by Ray. With the current implementation, Ray will try to pack actors on nodes
until the resource utilization is beyond a certain threshold and spread actors afterwards.

> NodeAffinitySchedulingStrategy is a low-level strategy that allows an actor to be scheduled onto a particular node specified by its hexadecimal node id.

Reviewer: let's omit the `hexadecimal` here -- from the user's perspective, there should be only one node id and they shouldn't need to worry if it is hexadecimal or binary etc.

Scheduling Strategy
-------------------

> Actors support ``scheduling_strategy`` option to specify the strategy used to decide the best node among available nodes.

Reviewer 2: ```suggestion
Actors support a ``scheduling_strategy`` option to specify the strategy used to decide the best node among available nodes.
```

> Currently the supported strategy for actors are ``"DEFAULT"`` and ``ray.util.scheduling_strategies.NodeSchedulingStrategy(node_id, soft)``.

Reviewer 2: ```suggestion
Currently the supported strategies for actors are ``"DEFAULT"`` and ``ray.util.scheduling_strategies.NodeSchedulingStrategy(node_id, soft)``.
```

``"DEFAULT"`` is the default strategy used by Ray. With the current implementation, Ray will try to pack actors on nodes
until the resource utilization is beyond a certain threshold and spread actors afterwards.

> NodeSchedulingStrategy is the low level strategy that allows an actor to be scheduled onto a particular node specified by the hex node id.

Reviewer 2: ```suggestion
NodeSchedulingStrategy is the low level strategy that allows an actor to be scheduled onto a particular node specified by its hexadecimal node id.
```

> The ``soft`` flag specifies whether the actor is allowed to run somewhere else if the specified node doesn't exist (e.g. the node dies)

Reviewer 2: ```suggestion
The ``soft`` flag specifies whether the actor is allowed to run somewhere else if the specified node doesn't exist (e.g., if the node dies)
```

or is infeasible for the required resources. If it is True, the actor will be scheduled onto other feasible nodes under such case.

> If it's False, the actor will fail with ``ActorUnschedulableError``.

Reviewer 2: Love this.

        # Zero-CPU (and no other resources) actors are randomly assigned to nodes.
        a2 = Actor.options(num_cpus=0).remote()

        # Only run the actor on the local node.
        a3 = Actor.options(
            scheduling_strategy=NodeAffinitySchedulingStrategy(

>                 node_id = ray.get_runtime_context().node_id.hex(),

Reviewer: Before merging this, we wanted to have a standardized example that doesn't need the `.hex()` method here, right? Actually maybe the cleanest solution here is to just omit the `.hex()` here and have `node_id` be of type `NodeID` -- this would parallel our treatment of `ObjectRef` as a Python type.

Reviewer 3: Hmm I agree with you, but we decided on string types for simplicity (see [above conversation](https://github.com/ray-project/ray/pull/23381#issuecomment-1085387958)).

It would be easy to keep backwards compatibility for string types vs NodeIDs, but I agree compatibility will be more annoying for the `.hex()` call so it's more straightforward for users to just use whatever type is returned by the existing ray APIs. What do you think, @ericl?

Author: We do want to change runtime_context to return node_id as hex string: https://github.com/ray-project/ray/pull/23715 but that's backward incompatible and people think we should do it for Ray 2.0. 

What I can do here is changing NodeAffinitySchedulingStrategy to accept both NodeID and string so we won't see `.hex()` in the user doc and in Ray 2.0 drop the support for NodeID. WDYT?

Reviewer 3: Ah didn't realize we were going to drop user-facing NodeID anyway. In that case I like @jjyao's plan to accept both for now so it's easy to deprecate NodeID later!

Reviewer: I'm fine with using a string, but then the user should only ever see the hex version of the string and there should be no other public APIs the return NodeID or the binary node id.

This is the first time we introduce a public API dependency on the node id, so I think we should do it as part of this PR and not introduce this `.hex()` thing now and then change it later for Ray 2.0. Just be very precise what the public API is on getting node ids and how they can be passed to `NodeAffinitySchedulingStrategy`.

Author: The issue is that we already have a (and the only) public api that returns NodeID (i.g. runtime_context) and I think we shouldn't change that API in this PR.

With my proposal, users don't need to call `.hex()`, they can directly use the node id returned by runtime_context. Later when we change runtime_context to return string instead of NodeID, the same code will still work so it's transparent to the users.

We can also make this api private (no user-facing document) for now until we have public apis to get string node id.

Reviewer: The plan to accept both for now and then migrate for Ray 2.0 sounds good to me!
